<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
<meta http-equiv="x-ua-compatible" content="IE=9" >

<title>Notes on Computational Genomics with R</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>





</head>

<body>
<h1>Notes on Computational Genomics with R</h1>

<h4>by Altuna Akalin</h4>

<p>The aim of computational genomics is to do biological interpretation of
 high dimensional genomics data.
 Generally speaking, it is similar to any other kind of data analysis but
 often times doing computational genomics will require domain specific knowledge
 and tools.
 The tasks of computational genomics can be roughly summarized as in Figure</p>

<p>Most of the time, the analysis starts with the raw data (if you are somehow
 served with already processed data, consider yourself lucky).
 The raw data could be image files from a microarray, or text files from
 a sequencer.
 That means the analysis starts with processing the data to a manageable
 format.
 This processing includes data munging which is transforming the data from
 one format to another, data point transformations such as normalizing and
 log-transformations, and filtering and removing data points with specific
 thresholds.
 For example, you may want to remove data points that have unreliable measuremen
ts or missing values.
 The next step is to apply supervised/unsupervised learning algorithms to
 test the hypothesis that lead to experiments that generated the data.
 This step also includes quality checks about your data.
 Another important thing to do is to annotate and visualize your data.
 For example, after finding transcription factor binding sites using data
 from ChIP-seq experiments, you would like to see what kind of genes they
 are nearby or what kind of other genomic features they overlap with.
 Do they overlap with promoters or they are distal? You may also like to
 see your binding sites on the genome or make a summary plot showing distances
 to nearest transcriptions start sites.
 All of this tasks can be classified as visualization and annotation tasks.
 All of these steps should hopefully lead you to the holy grail which is
 biological interpretation of the data and hopefully to some new insights
 about genome biology.</p>

</body>

</html>

