<!DOCTYPE HTML>
<html lang="en-US">
    <head prefix="og: http://ogp.me/ns# book: http://ogp.me/ns/book#">
        
        <meta charset="UTF-8">
        <title>The Setup | Notes on Computational Genomics with R</title>

        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="robots" content="index, follow">
        <meta name="author" content="al2na">
        <meta name="description" content="This is somewhat an opinionated guide on using R for computational genomics. It is aimed at wet-lab researchers who wants to use R in their data analysis ,and bioinformaticians who are new to R and wants to learn more about its capabilities for genomics data analysis.">
        <meta name="keywords" content="gitbook,github" >
        <meta name="generator" content="www.gitbook.io">

        
        <link rel="next" href="../intro_to_r/computations_in_r.html" />
        
        
        <link rel="prev" href="../intro_to_r/README.html" />
        

        <meta property="og:title" content="The Setup | Notes on Computational Genomics with R">
        <meta property="og:site_name" content="Notes on Computational Genomics with R">
        <meta property="og:type" content="book">
        <meta property="og:locale" content="en_US">

        <meta property="book:author" content="https://github.com/al2na">
        <meta property="book:tag" content="GitBook">

        <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
        <meta name="apple-mobile-web-app-capable" content="yes">
        <meta name="apple-mobile-web-app-status-bar-style" content="black">

        <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">
        
    </head>
    <body>
        
        <link rel="stylesheet" href="../gitbook/style.css">
        
        
    <div class="book" data-github="al2na/compgenr" data-level="2.1" data-basepath=".." data-revision="1401401823619">
    <div class="book-header">
    <!-- Actions Left -->
    
    <a href="https://github.com/al2na/compgenr" target="_blank" class="btn pull-left"><i class="fa fa-github-alt"></i></a>
    
    <a href="#" class="btn pull-left toggle-summary"><i class="fa fa-align-justify"></i></a>
    <a href="#" class="btn pull-left toggle-search"><i class="fa fa-search"></i></a>

    <!-- Actions Right -->
    <a href="#" target="_blank" class="btn pull-right" data-sharing="google-plus"><i class="fa fa-google-plus"></i></a>
    <a href="#" target="_blank" class="btn pull-right" data-sharing="facebook"><i class="fa fa-facebook"></i></a>
    <a href="#" target="_blank" class="btn pull-right" data-sharing="twitter"><i class="fa fa-twitter"></i></a>

    
    <a href="https://github.com/al2na/compgenr/stargazers" target="_blank" class="btn pull-right count-star"><i class="fa fa-star-o"></i> Star (<span>-</span>)</a>
    <a href="https://github.com/al2na/compgenr/watchers" target="_blank" class="btn pull-right count-watch"><i class="fa fa-eye"></i> Watch (<span>-</span>)</a>
    

    <!-- Title -->
    <h1><a href="../" >Notes on Computational Genomics with R</a></h1>
</div>

    <div class="book-summary">
    <div class="book-search">
        <input type="text" placeholder="Search" class="form-control" />
    </div>
    <ul class="summary">
        
        <li>
            <a href="https://github.com/al2na" target="blank">About the author</a>
        </li>
        <li>
            <a href="https://github.com/al2na/compgenr/issues" target="blank">Questions and Issues</a>
        </li>
        <li>
            <a href="https://github.com/al2na/compgenr/edit/master/intro_to_r/the_setup.md" target="blank">Edit and Contribute</a>
        </li>
        <li class="divider"></li>
        
        <li data-level="0" data-path="index.html">
            <a href="../index.html"><i class="fa fa-check"></i> Introduction</a>
        </li>
        
            <li  data-level="1" data-path="compgenr/README.html">
                
                <a href="../compgenr/README.html">
                    <i class="fa fa-check"></i> <b>1.</b> Computational Genomics and R
                </a>
                
                
                <ul class="articles">
                    
                        <li  data-level="1.1" data-path="compgenr/what_can_you_do_with_r.html">
                            
                            <a href="../compgenr/what_can_you_do_with_r.html">
                                <i class="fa fa-check"></i> <b>1.1.</b> What can you do with R ?
                            </a>
                            
                        </li>
                    
                </ul>
                
            </li>
        
            <li  data-level="2" data-path="intro_to_r/README.html">
                
                <a href="../intro_to_r/README.html">
                    <i class="fa fa-check"></i> <b>2.</b> Quick Introduction to R
                </a>
                
                
                <ul class="articles">
                    
                        <li  data-level="2.1" data-path="intro_to_r/the_setup.html">
                            
                            <a href="../intro_to_r/the_setup.html">
                                <i class="fa fa-check"></i> <b>2.1.</b> The Setup
                            </a>
                            
                        </li>
                    
                        <li  data-level="2.2" data-path="intro_to_r/computations_in_r.html">
                            
                            <a href="../intro_to_r/computations_in_r.html">
                                <i class="fa fa-check"></i> <b>2.2.</b> Computations in R
                            </a>
                            
                        </li>
                    
                        <li  data-level="2.3" data-path="intro_to_r/data_types.html">
                            
                            <a href="../intro_to_r/data_types.html">
                                <i class="fa fa-check"></i> <b>2.3.</b> Data structures and  types
                            </a>
                            
                        </li>
                    
                        <li  data-level="2.4" data-path="intro_to_r/reading_and_writing_data.html">
                            
                            <a href="../intro_to_r/reading_and_writing_data.html">
                                <i class="fa fa-check"></i> <b>2.4.</b> Reading and writing data
                            </a>
                            
                        </li>
                    
                        <li  data-level="2.5" data-path="intro_to_r/plotting_in_r.html">
                            
                            <a href="../intro_to_r/plotting_in_r.html">
                                <i class="fa fa-check"></i> <b>2.5.</b> Plotting in R
                            </a>
                            
                        </li>
                    
                        <li  data-level="2.6" data-path="intro_to_r/functions_and_control_structures_for,_ifelse_etc.html">
                            
                            <a href="../intro_to_r/functions_and_control_structures_for,_ifelse_etc.html">
                                <i class="fa fa-check"></i> <b>2.6.</b> Functions and control structures (for, if/else etc.) 
                            </a>
                            
                        </li>
                    
                </ul>
                
            </li>
        
            <li  data-level="3" data-path="genomic_intervals/README.html">
                
                <a href="../genomic_intervals/README.html">
                    <i class="fa fa-check"></i> <b>3.</b> Genomic Intervals
                </a>
                
                
                <ul class="articles">
                    
                        <li  data-level="3.1" data-path="genomic_intervals/granges.html">
                            
                            <a href="../genomic_intervals/granges.html">
                                <i class="fa fa-check"></i> <b>3.1.</b> Operations on Genomic Intervals with GenomicRanges package
                            </a>
                            
                        </li>
                    
                        <li  data-level="3.2" data-path="genomic_intervals/htseq.html">
                            
                            <a href="../genomic_intervals/htseq.html">
                                <i class="fa fa-check"></i> <b>3.2.</b> Dealing with high-throughput sequencing reads
                            </a>
                            
                        </li>
                    
                        <li  data-level="3.3" data-path="genomic_intervals/scores.html">
                            
                            <a href="../genomic_intervals/scores.html">
                                <i class="fa fa-check"></i> <b>3.3.</b> Dealing with contiguous scores over the genome
                            </a>
                            
                        </li>
                    
                </ul>
                
            </li>
        
            <li  data-level="4" data-path="sequence_analysis/README.html">
                
                <a href="../sequence_analysis/README.html">
                    <i class="fa fa-check"></i> <b>4.</b> Sequence analysis
                </a>
                
                
            </li>
        
            <li  data-level="5" data-path="visualization/README.html">
                
                <a href="../visualization/README.html">
                    <i class="fa fa-check"></i> <b>5.</b> Visualization
                </a>
                
                
            </li>
        
    </ul>
</div>

    <div class="book-body">
        <div class="body-inner">
            <div class="page-wrapper" tabindex="-1">
                <div class="book-progress">
    <div class="bar">
        <div class="inner" style="width: 26.666666666666668%;min-width: 20%;"></div>
    </div>
    <div class="chapters">
    
        <a href="../index.html" title="Introduction" class="chapter done new-chapter" data-progress="0" style="left: 0%;"></a>
    
        <a href="../compgenr/README.html" title="Computational Genomics and R" class="chapter done new-chapter" data-progress="1" style="left: 6.666666666666667%;"></a>
    
        <a href="../compgenr/what_can_you_do_with_r.html" title="What can you do with R ?" class="chapter done " data-progress="1.1" style="left: 13.333333333333334%;"></a>
    
        <a href="../intro_to_r/README.html" title="Quick Introduction to R" class="chapter done new-chapter" data-progress="2" style="left: 20%;"></a>
    
        <a href="../intro_to_r/the_setup.html" title="The Setup" class="chapter done " data-progress="2.1" style="left: 26.666666666666668%;"></a>
    
        <a href="../intro_to_r/computations_in_r.html" title="Computations in R" class="chapter  " data-progress="2.2" style="left: 33.333333333333336%;"></a>
    
        <a href="../intro_to_r/data_types.html" title="Data structures and  types" class="chapter  " data-progress="2.3" style="left: 40%;"></a>
    
        <a href="../intro_to_r/reading_and_writing_data.html" title="Reading and writing data" class="chapter  " data-progress="2.4" style="left: 46.666666666666664%;"></a>
    
        <a href="../intro_to_r/plotting_in_r.html" title="Plotting in R" class="chapter  " data-progress="2.5" style="left: 53.333333333333336%;"></a>
    
        <a href="../intro_to_r/functions_and_control_structures_for,_ifelse_etc.html" title="Functions and control structures (for, if/else etc.) " class="chapter  " data-progress="2.6" style="left: 60%;"></a>
    
        <a href="../genomic_intervals/README.html" title="Genomic Intervals" class="chapter  new-chapter" data-progress="3" style="left: 66.66666666666667%;"></a>
    
        <a href="../genomic_intervals/granges.html" title="Operations on Genomic Intervals with GenomicRanges package" class="chapter  " data-progress="3.1" style="left: 73.33333333333333%;"></a>
    
        <a href="../genomic_intervals/htseq.html" title="Dealing with high-throughput sequencing reads" class="chapter  " data-progress="3.2" style="left: 80%;"></a>
    
        <a href="../genomic_intervals/scores.html" title="Dealing with contiguous scores over the genome" class="chapter  " data-progress="3.3" style="left: 86.66666666666667%;"></a>
    
        <a href="../sequence_analysis/README.html" title="Sequence analysis" class="chapter  new-chapter" data-progress="4" style="left: 93.33333333333333%;"></a>
    
        <a href="../visualization/README.html" title="Visualization" class="chapter  new-chapter" data-progress="5" style="left: 100%;"></a>
    
    </div>
</div>

                <div class="page-inner">
                
                    <section class="normal" id="section-gitbook_14">
                    
                        <h1 id="the-setup">The Setup</h1>
<p>Download and install R <a href="http://cran.r-project.org/" target="_blank">http://cran.r-project.org/</a> and RStudio <a href="http://www.rstudio.com/" target="_blank">http://www.rstudio.com/</a> if you do not have them already. Rstudio is optional but it is a great tool if you are just starting to learn R.
You will need specific data sets to run the codes in this document. Download the data.zip[URL to come] and extract it to your directory of choice. The folder name should be “data” and your R working directory should be level above the data folder. That means in your R console, when you type “dir(“data”)” you should be able to see the contents of the data folder. You can change your working directory by <em>setwd()</em> command and get your current working directory with <em>getwd()</em> command in R2. In RStudio, you can click on the top menu and change the location of your working directory via user interface.</p>
<h2 id="installing-packages">Installing packages</h2>
<p>R packages are add-ons to base R that help you achieve additional tasks that are not directly supported by base R. It is by the action of these extra functionality that R excels as a tool for computational genomics. Bioconductor project (<a href="http://bioconductor.org/" target="_blank">http://bioconductor.org/</a>) is a dedicated package repository for computational biology related packages. However main package repository of R, called CRAN, has also computational biology related packages. In addition, R-Forge(<a href="http://r-forge.r-project.org/" target="_blank">http://r-forge.r-project.org/</a>), GitHub(<a href="https://github/" target="_blank">https://github</a>. com/), and googlecode(<a href="http://code.google.com/" target="_blank">http://code.google.com</a>) are other locations where R packages might be hosted.
You can install CRAN packages using install.packages(). (# is the comment character in R)</p>
<pre><code class="lang-r"><span class="hljs-comment"># install package named 'randomForests' from CRAN</span>
install.packages(<span class="hljs-string">"randomForests"</span>)
</code></pre>
<p>You can install bioconductor packages with a specific installer script</p>
<pre><code class="lang-r"><span class="hljs-comment"># get the installer package</span>
<span class="hljs-keyword">source</span>(<span class="hljs-string">"http://bioconductor.org/biocLite.R"</span>)
<span class="hljs-comment"># install bioconductor package 'rtracklayer'</span>
biocLite(<span class="hljs-string">"rtracklayer"</span>)
</code></pre>
<p>You can install packages from github using <em>install_github()</em> function from <strong>devtools</strong></p>
<pre><code class="lang-r"><span class="hljs-keyword">library</span>(devtools)
install_github(<span class="hljs-string">"roxygen"</span>)
</code></pre>
<p>Another way to install packages are from the source.</p>
<pre><code class="lang-r"><span class="hljs-comment"># download the source file</span>
download.file(<span class="hljs-string">"http://goo.gl/3pvHYI"</span>, destfile = <span class="hljs-string">"methylKit_0.5.7.tar.gz"</span>)
<span class="hljs-comment"># install the package from the source file</span>
install.packages(<span class="hljs-string">"methylKit_0.5.7.tar.gz"</span>, repos = <span class="hljs-literal">NULL</span>, type = <span class="hljs-string">"source"</span>)
<span class="hljs-comment"># delete the source file</span>
unlink(<span class="hljs-string">"methylKit_0.5.7.tar.gz"</span>)
</code></pre>
<p>You  can also update CRAN and Bioconductor packages.</p>
<pre><code class="lang-r"><span class="hljs-comment"># updating CRAN packages</span>
update.packages()

<span class="hljs-comment"># updating bioconductor packages</span>
<span class="hljs-keyword">source</span>(<span class="hljs-string">"http://bioconductor.org/biocLite.R"</span>)
biocLite(<span class="hljs-string">"BiocUpgrade"</span>)
</code></pre>
<h2 id="installing-packages-in-custom-locations">Installing packages in custom locations</h2>
<p>If you will be using R on servers or computing clusters rather than your personal computer it is unlikey that you will have administrator access to install packages. In that case, you can install packges in custom locations by telling R where to look for additional packages. This is done by setting up an .Renviron file in your home directory and add the following line:</p>
<pre><code>R_LIBS=~/Rlibs
</code></pre><p>This tells R that “Rlibs” directory at your home directory will be the first choice of locations to look for packages and install packages (The directory name and location is up to you above is just an example). You should go and create that directory now. After that, start a fresh R session and start installing packages. From now on, packages will be installed to your local directory where you have read-write access.</p>
<h2 id="getting-help-on-functions-and-packages">Getting help on functions and packages</h2>
<p>You can get help on functions by help() and help.search() functions. You can list the functions in a package with ls() function</p>
<pre><code class="lang-r">libray(MASS)
ls(<span class="hljs-string">"package:MASS"</span>) <span class="hljs-comment"># functions in the package</span>
ls() <span class="hljs-comment"># objects in your R enviroment</span>
<span class="hljs-comment"># get help on hist() function</span>
?hist
help(<span class="hljs-string">"hist"</span>)
<span class="hljs-comment"># search the word "hist" in help pages</span>
help.search(<span class="hljs-string">"hist"</span>)
??hist
@
</code></pre>
<h3 id="more-help-needed-">More help needed?</h3>
<p>In addition, check package vignettes for help and practical understanding of the functions. All Bionconductor packages have vignettes that walk you thorugh example analysis. Google search will always be helpful as well, there are many blogs and web pages that have posts about R. R-help, Stackoverflow and R-bloggers are usually source of good and reliable information.</p>

                    
                    </section>
                
                </div>
            </div>
        </div>

        
        <a href="../intro_to_r/README.html" class="navigation navigation-prev"><i class="fa fa-angle-left"></i></a>
        
        
        <a href="../intro_to_r/computations_in_r.html" class="navigation navigation-next"><i class="fa fa-angle-right"></i></a>
        
    </div>
</div>

        
        <script type="text/javascript">(function(e,b){if(!b.__SV){var a,f,i,g;window.mixpanel=b;b._i=[];b.init=function(a,e,d){function f(b,h){var a=h.split(".");2==a.length&&(b=b[a[0]],h=a[1]);b[h]=function(){b.push([h].concat(Array.prototype.slice.call(arguments,0)))}}var c=b;"undefined"!==typeof d?c=b[d]=[]:d="mixpanel";c.people=c.people||[];c.toString=function(b){var a="mixpanel";"mixpanel"!==d&&(a+="."+d);b||(a+=" (stub)");return a};c.people.toString=function(){return c.toString(1)+".people (stub)"};i="disable track track_pageview track_links track_forms register register_once alias unregister identify name_tag set_config people.set people.set_once people.increment people.append people.track_charge people.clear_charges people.delete_user".split(" ");
for(g=0;g<i.length;g++)f(c,i[g]);b._i.push([a,e,d])};b.__SV=1.2;a=e.createElement("script");a.type="text/javascript";a.async=!0;a.src=("https:"===e.location.protocol?"https:":"http:")+'//cdn.mxpnl.com/libs/mixpanel-2.2.min.js';f=e.getElementsByTagName("script")[0];f.parentNode.insertBefore(a,f)}})(document,window.mixpanel||[]);
        </script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/ace/1.1.3/ace.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/ace/1.1.3/mode-javascript.js"></script>
        <script src="../gitbook/jsrepl/jsrepl.js" id="jsrepl-script"></script>
        <script src="../gitbook/app.js"></script>
        
    </body>
</html>
