<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
</head>
<body>
<p>@path/Users/altuna/Dropbox/PAPERS/R-devel/compgenr/chapters//</p>
<h1 id="computational-genomics-and-r">Computational genomics and R</h1>
<h2 id="computational-genomics">Computational Genomics</h2>
<p>The aim of computational genomics is to do biological interpretation of high dimensional genomics data. Generally speaking, it is similar to any other kind of data analysis but often times doing computational genomics will require domain specific knowledge and tools. The tasks of computational genomics can be roughly summarized as in Figure [fig:Computational-genomics-(rough)].</p>
<div class="figure">
<embed src="0_Users_altuna_Dropbox_PAPERS_R-devel_compgenr_chapters_nonR_figures_dataAnalysis.pdf" /><p class="caption">Computational genomics (rough) flowchart[fig:Computational-genomics-(rough)]</p>
</div>
<p>Most of the time, the analysis starts with the raw data (if you are somehow served with already processed data, consider yourself lucky). The raw data could be image files from a microarray, or text files from a sequencer. That means the analysis starts with processing the data to a manageable format. This processing includes data munging which is transforming the data from one format to another, data point transformations such as normalizing and log-transformations, and filtering and removing data points with specific thresholds. For example, you may want to remove data points that have unreliable measurements or missing values. The next step is to apply supervised/unsupervised learning algorithms to test the hypothesis that lead to experiments that generated the data. This step also includes quality checks about your data. Another important thing to do is to annotate and visualize your data. For example, after finding transcription factor binding sites using data from ChIP-seq experiments, you would like to see what kind of genes they are nearby or what kind of other genomic features they overlap with. Do they overlap with promoters or they are distal? You may also like to see your binding sites on the genome or make a summary plot showing distances to nearest transcriptions start sites. All of this tasks can be classified as visualization and annotation tasks. All of these steps should hopefully lead you to the holy grail which is biological interpretation of the data and hopefully to some new insights about genome biology.</p>
<h2 id="what-can-you-do-with-r">What can you do with R?</h2>
<p>The tasks that are descrived above can be accomplished by R. R is not only a powerful statistical programming language but also go-to data analysis tool for many computational genomics experts. High-dimensional genomics datasets are usually suitable to be analyzed with core R packages and functions. On top of that, Bioconductor and CRAN have an array of specialized tools for doing genomics specific analysis.</p>
<p>Here is a list of computational genomics tasks that can be completed using R.</p>
<h3 id="data-munging-pre-processing">Data munging (pre-processing)</h3>
<p>Often times, the data will not come in ready to analyze format. You may need to convert it to other formats by transforming data points (such as log transforming, normalizing etc), or remove columns/rows that , or remove data points with empty values and, and subset the data set with some arbitrary condition. Most of these tasks can be achieved using R. In addition, with the help of packages R can connect to databases in various formats such as mySQL, mongoDB, etc., and query and get the data to R environment using database specific tools. Unfortunately, not all data muging and processing tasks can be accomplished only by R. At times, you may need to use domain specific software or software dealing better with specific type of data sets. For example, R is not great at dealing with character strings, if you are trying to filter a large dataset based on some regular expression you may be better of with perl or awk.</p>
<h3 id="general-data-analysis-and-exploration">General data analysis and exploration</h3>
<p>Most genomics data sets are suitable for application of general data analysis tools. In some cases, you may need to preprocess the data to get it to a state that is suitable for application such tools.</p>
<ul>
<li><p>unsupervised data analysis: clustering (k-means, hierarchical), matrix factorization (PCA, ICA etc)</p></li>
<li><p>supervised data analysis: generalized linear models, support vector machines, randomForests</p></li>
</ul>
<h3 id="visualization">Visualization</h3>
<p>Visualization is an important part of all data analysis techniques including computational genomics. Again, you can use core visualization technniques in R and also genomics specific ones with the help of specific packages.</p>
<ul>
<li><p>Basic plots: Histograms, scatter plots, bar plots, box plots</p></li>
<li><p>ideograms and circus plots for genomics</p></li>
<li><p>heatmaps</p></li>
<li><p>meta-profiles of genomic features, read enrichment over all promoters</p></li>
<li><p>genomic track visualization for given locus</p></li>
</ul>
<h3 id="dealing-with-genomic-intervals">Dealing with genomic intervals</h3>
<p>Most of the genomics data come in a tabular format that contains the location in the genome and some other relevant values, such as scores for those genomic features and/or names. R/Bioconductor has dedicated methods to deal with such data. Here are a couple of example tasks that you can achieve using R.</p>
<ul>
<li><p>Overlapping CpG islands with transcription start sites, and filtering based on overlaps</p></li>
<li><p>Aligning reads and making read enrichment profiles</p></li>
<li><p>Overlapping aligned reads with exons and counting aligned reads per gene</p></li>
</ul>
<h3 id="application-of-other-bioinformatics-specific-algorithms">Application of other bioinformatics specific algorithms</h3>
<p>In addition to genomic interval centered methods, R/Bioconductor gives you access to multitude of other bioinformatics specific algorithms. Here are some of the things you can do.</p>
<ul>
<li><p>Sequence analysis: TF binding motifs, GC content and CpG counts of a given DNA sequence</p></li>
<li><p>Differential expression (or arrays and sequencing based measurements)</p></li>
<li><p>Gene set/Pathway analysis: What kind of genes are enriched in my gene set</p></li>
</ul>
</body>
</html>
