\batchmode
\makeatletter
\def\input@path{{/Users/altuna/Dropbox/PAPERS/R-devel/compgenr/chapters//}}
\makeatother
\documentclass[english,nohyper]{tufte-book}\usepackage[]{graphicx}\usepackage[]{color}
%% maxwidth is the original width if it is less than linewidth
%% otherwise use linewidth (to make sure the graphics do not exceed the margin)
\makeatletter
\def\maxwidth{ %
  \ifdim\Gin@nat@width>\linewidth
    \linewidth
  \else
    \Gin@nat@width
  \fi
}
\makeatother

\definecolor{fgcolor}{rgb}{0.345, 0.345, 0.345}
\newcommand{\hlnum}[1]{\textcolor[rgb]{0.686,0.059,0.569}{#1}}%
\newcommand{\hlstr}[1]{\textcolor[rgb]{0.192,0.494,0.8}{#1}}%
\newcommand{\hlcom}[1]{\textcolor[rgb]{0.678,0.584,0.686}{\textit{#1}}}%
\newcommand{\hlopt}[1]{\textcolor[rgb]{0,0,0}{#1}}%
\newcommand{\hlstd}[1]{\textcolor[rgb]{0.345,0.345,0.345}{#1}}%
\newcommand{\hlkwa}[1]{\textcolor[rgb]{0.161,0.373,0.58}{\textbf{#1}}}%
\newcommand{\hlkwb}[1]{\textcolor[rgb]{0.69,0.353,0.396}{#1}}%
\newcommand{\hlkwc}[1]{\textcolor[rgb]{0.333,0.667,0.333}{#1}}%
\newcommand{\hlkwd}[1]{\textcolor[rgb]{0.737,0.353,0.396}{\textbf{#1}}}%

\usepackage{framed}
\makeatletter
\newenvironment{kframe}{%
 \def\at@end@of@kframe{}%
 \ifinner\ifhmode%
  \def\at@end@of@kframe{\end{minipage}}%
  \begin{minipage}{\columnwidth}%
 \fi\fi%
 \def\FrameCommand##1{\hskip\@totalleftmargin \hskip-\fboxsep
 \colorbox{shadecolor}{##1}\hskip-\fboxsep
     % There is no \\@totalrightmargin, so:
     \hskip-\linewidth \hskip-\@totalleftmargin \hskip\columnwidth}%
 \MakeFramed {\advance\hsize-\width
   \@totalleftmargin\z@ \linewidth\hsize
   \@setminipage}}%
 {\par\unskip\endMakeFramed%
 \at@end@of@kframe}
\makeatother

\definecolor{shadecolor}{rgb}{.97, .97, .97}
\definecolor{messagecolor}{rgb}{0, 0, 0}
\definecolor{warningcolor}{rgb}{1, 0, 1}
\definecolor{errorcolor}{rgb}{1, 0, 0}
\newenvironment{knitrout}{}{} % an empty environment to be redefined in TeX

\usepackage{alltt}
\usepackage[T1]{fontenc}
\usepackage[latin9]{inputenc}
\usepackage{graphicx}

\makeatletter
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% User specified LaTeX commands.
\usepackage{epigraph}

\makeatother

\usepackage{babel}
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\begin{document}






\chapter{Computational genomics and R}


\section{Computational Genomics}

The aim of computational genomics is to do biological interpretation
of high dimensional genomics data. Generally speaking, it is similar
to any other kind of data analysis but often times doing computational
genomics will require domain specific knowledge and tools. The tasks
of computational genomics can be roughly summarized as in Figure \ref{fig:Computational-genomics-(rough)}. 

\begin{figure}
\includegraphics[scale=0.5]{0_Users_altuna_Dropbox_PAPERS_R-devel_compgenr_chapters_nonR_figures_dataAnalysis.pdf}\caption{Computational genomics (rough) flowchart\label{fig:Computational-genomics-(rough)}}


\end{figure}


Most of the time, the analysis starts with the raw data (if you are
somehow served with already processed data, consider yourself lucky).
The raw data could be image files from a microarray, or text files
from a sequencer. That means the analysis starts with processing the
data to a manageable format. This processing includes data munging
which is transforming the data from one format to another, data point
transformations such as normalizing and log-transformations, and filtering
and removing data points with specific thresholds. For example, you
may want to remove data points that have unreliable measurements or
missing values. The next step is to apply supervised/unsupervised
learning algorithms to test the hypothesis that lead to experiments
that generated the data. This step also includes quality checks about
your data. Another important thing to do is to annotate and visualize
your data. For example, after finding transcription factor binding
sites using data from ChIP-seq experiments, you would like to see
what kind of genes they are nearby or what kind of other genomic features
they overlap with. Do they overlap with promoters or they are distal?
You may also like to see your binding sites on the genome or make
a summary plot showing distances to nearest transcriptions start sites.
All of this tasks can be classified as visualization and annotation
tasks. All of these steps should hopefully lead you to the holy grail
which is biological interpretation of the data and hopefully to some
new insights about genome biology.


\section{What can you do with R?}

The tasks that are descrived above can be accomplished by R. R is
not only a powerful statistical programming language but also go-to
data analysis tool for many computational genomics experts. High-dimensional
genomics datasets are usually suitable to be analyzed with core R
packages and functions. On top of that, Bioconductor and CRAN have
an array of specialized tools for doing genomics specific analysis. 

Here is a list of computational genomics tasks that can be completed
using R.


\subsection{Data munging (pre-processing)}

Often times, the data will not come in ready to analyze format. You
may need to convert it to other formats by transforming data points
(such as log transforming, normalizing etc), or remove columns/rows
that , or remove data points with empty values and, and subset the
data set with some arbitrary condition. Most of these tasks can be
achieved using R. In addition, with the help of packages R can connect
to databases in various formats such as mySQL, mongoDB, etc., and
query and get the data to R environment using database specific tools.
Unfortunately, not all data muging and processing tasks can be accomplished
only by R. At times, you may need to use domain specific software
or software dealing better with specific type of data sets. For example,
R is not great at dealing with character strings, if you are trying
to filter a large dataset based on some regular expression you may
be better of with perl or awk. 


\subsection{General data analysis and exploration}

Most genomics data sets are suitable for application of general data
analysis tools. In some cases, you may need to preprocess the data
to get it to a state that is suitable for application such tools.
\begin{itemize}
\item unsupervised data analysis: clustering (k-means, hierarchical), matrix
factorization (PCA, ICA etc)
\item supervised data analysis: generalized linear models, support vector
machines, randomForests
\end{itemize}

\subsection{Visualization}

Visualization is an important part of all data analysis techniques
including computational genomics. Again, you can use core visualization
technniques in R and also genomics specific ones with the help of
specific packages.
\begin{itemize}
\item Basic plots: Histograms, scatter plots, bar plots, box plots
\item ideograms and circus plots for genomics
\item heatmaps
\item meta-profiles of genomic features, read enrichment over all promoters
\item genomic track visualization for given locus
\end{itemize}

\subsection{Dealing with genomic intervals}

Most of the genomics data come in a tabular format that contains the
location in the genome and some other relevant values, such as scores
for those genomic features and/or names. R/Bioconductor has dedicated
methods to deal with such data. Here are a couple of example tasks
that you can achieve using R.
\begin{itemize}
\item Overlapping CpG islands with transcription start sites, and filtering
based on overlaps
\item Aligning reads and making read enrichment profiles
\item Overlapping aligned reads with exons and counting aligned reads per
gene
\end{itemize}

\subsection{Application of other bioinformatics specific algorithms}

In addition to genomic interval centered methods, R/Bioconductor gives
you access to multitude of other bioinformatics specific algorithms.
Here are some of the things you can do. 
\begin{itemize}
\item Sequence analysis: TF binding motifs, GC content and CpG counts of
a given DNA sequence
\item Differential expression (or arrays and sequencing based measurements)
\item Gene set/Pathway analysis: What kind of genes are enriched in my gene
set\end{itemize}

\end{document}
